File: rewrite-javascript/rewrite/src/json/visitor.ts
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {mapAsync} from "../util";
import {produceAsync, TreeVisitor, ValidImmerRecipeReturnType} from "../visitor";
import {SourceFile} from "../tree";
import {isJson, Json} from "./tree";
import {createDraft, Draft, finishDraft} from "immer";

export class JsonVisitor<P> extends TreeVisitor<Json, P> {
    async isAcceptable(sourceFile: SourceFile): Promise<boolean> {
        return isJson(sourceFile);
    }

    protected async visitArray(array: Json.Array, p: P): Promise<Json | undefined> {
        return this.produceJson<Json.Array>(array, p, async draft => {
            draft.values = await mapAsync(array.values, value => this.visitRightPadded(value, p));
        })
    }

    protected async visitDocument(document: Json.Document, p: P): Promise<Json | undefined> {
        return this.produceJson<Json.Document>(document, p, async draft => {
            draft.value = await this.visitDefined(document.value, p);
            draft.eof = await this.visitSpace(document.eof, p);
        })
    }

    protected async visitEmpty(empty: Json.Empty, p: P): Promise<Json | undefined> {
        return this.produceJson(empty, p)
    }

    protected async visitIdentifier(identifier: Json.Identifier, p: P): Promise<Json | undefined> {
        return this.produceJson(identifier, p)
    }

    protected async visitLiteral(literal: Json.Literal, p: P): Promise<Json | undefined> {
        return this.produceJson(literal, p)
    }

    protected async visitMember(member: Json.Member, p: P): Promise<Json | undefined> {
        return this.produceJson<Json.Member>(member, p, async draft => {
            draft.key = (await this.visitRightPadded(member.key, p))!;
            draft.value = await this.visitDefined(member.value, p);
        });
    }

    protected async visitObject(jsonObject: Json.Object, p: P): Promise<Json | undefined> {
        return this.produceJson<Json.Object>(jsonObject, p, async draft => {
            draft.members = await mapAsync(jsonObject.members, member => this.visitRightPadded(member, p));
        });
    }

    protected async visitRightPadded<T extends Json>(right: Json.RightPadded<T>, p: P):
        Promise<Json.RightPadded<T> | undefined> {
        return produceAsync<Json.RightPadded<T>>(right, async draft => {
            draft.element = await this.visitDefined(right.element, p);
            draft.after = await this.visitSpace(right.after, p)
        });
    }

    protected async visitSpace(space: Json.Space, p: P): Promise<Json.Space> {
        return space;
    }

    protected async produceJson<J extends Json>(
        before: Json | undefined,
        p: P,
        recipe?: (draft: Draft<J>) =>
            ValidImmerRecipeReturnType<Draft<J>> |
            PromiseLike<ValidImmerRecipeReturnType<Draft<J>>>
    ): Promise<J | undefined> {
        if (before === undefined) {
            return undefined;
        }
        const draft: Draft<J> = createDraft(before as J);
        (draft as Draft<Json>).prefix = await this.visitSpace(before!.prefix, p);
        (draft as Draft<Json>).markers = await this.visitMarkers(before!.markers, p);
        if (recipe) {
            await recipe(draft);
        }
        return finishDraft(draft) as J;
    }

    protected accept(t: Json, p: P): Promise<Json | undefined> {
        switch (t.kind) {
            case Json.Kind.Array:
                return this.visitArray(t as Json.Array, p);
            case Json.Kind.Document:
                return this.visitDocument(t as Json.Document, p);
            case Json.Kind.Empty:
                return this.visitEmpty(t as Json.Empty, p);
            case Json.Kind.Identifier:
                return this.visitIdentifier(t as Json.Identifier, p);
            case Json.Kind.Literal:
                return this.visitLiteral(t as Json.Literal, p);
            case Json.Kind.Member:
                return this.visitMember(t as Json.Member, p);
            case Json.Kind.Object:
                return this.visitObject(t as Json.Object, p);
            default:
                throw new Error(`Unexpected JSON kind ${t.kind}`);
        }
    }
}

File: rewrite-python/rewrite/src/rewrite/toml/visitor.py
from __future__ import annotations

from typing import TypeVar, Optional, Any

from rewrite.tree import Tree, SourceFile
from rewrite.visitor import TreeVisitor
from rewrite.toml.tree import (
    Document, Array, Table, KeyValue, Literal, Identifier, Empty,
    TomlSpace, TomlRightPadded
)

P = TypeVar('P')


class TomlVisitor(TreeVisitor[Tree, P]):

    def is_acceptable(self, source_file: SourceFile, p: P) -> bool:
        return isinstance(source_file, Document)

    def visit_document(self, document: Document, p: P) -> Document:
        d = document
        d = d.replace(prefix=self.visit_space(d.prefix, p))
        d = d.replace(markers=self.visit_markers(d.markers, p))
        new_values = []
        changed = False
        for v in d.values:
            result = self.visit(v, p)
            new_values.append(result)
            if result is not v:
                changed = True
        if changed:
            d = d.replace(values=new_values)
        d = d.replace(eof=self.visit_space(d.eof, p))
        return d

    def visit_array(self, array: Array, p: P) -> Array:
        a = array
        a = a.replace(prefix=self.visit_space(a.prefix, p))
        a = a.replace(markers=self.visit_markers(a.markers, p))
        new_values = []
        changed = False
        for v in a.values:
            result = self.visit_right_padded(v, p)
            new_values.append(result)
            if result is not v:
                changed = True
        if changed:
            a = a.replace(values=new_values)
        return a

    def visit_table(self, table: Table, p: P) -> Table:
        t = table
        t = t.replace(prefix=self.visit_space(t.prefix, p))
        t = t.replace(markers=self.visit_markers(t.markers, p))
        new_values = []
        changed = False
        for v in t.values:
            result = self.visit_right_padded(v, p)
            new_values.append(result)
            if result is not v:
                changed = True
        if changed:
            t = t.replace(values=new_values)
        return t

    def visit_key_value(self, key_value: KeyValue, p: P) -> KeyValue:
        kv = key_value
        kv = kv.replace(prefix=self.visit_space(kv.prefix, p))
        kv = kv.replace(markers=self.visit_markers(kv.markers, p))
        new_key = self.visit_right_padded(kv.key, p)
        if new_key is not kv.key:
            kv = kv.replace(key=new_key)
        new_value = self.visit(kv.value, p)
        if new_value is not kv.value:
            kv = kv.replace(value=new_value)
        return kv

    def visit_literal(self, literal: Literal, p: P) -> Literal:
        l = literal
        l = l.replace(prefix=self.visit_space(l.prefix, p))
        l = l.replace(markers=self.visit_markers(l.markers, p))
        return l

    def visit_identifier(self, identifier: Identifier, p: P) -> Identifier:
        i = identifier
        i = i.replace(prefix=self.visit_space(i.prefix, p))
        i = i.replace(markers=self.visit_markers(i.markers, p))
        return i

    def visit_empty(self, empty: Empty, p: P) -> Empty:
        e = empty
        e = e.replace(prefix=self.visit_space(e.prefix, p))
        e = e.replace(markers=self.visit_markers(e.markers, p))
        return e

    def visit_space(self, space: Optional[TomlSpace], p: P) -> Optional[TomlSpace]:
        return space

    def visit_right_padded(self, right: Optional[TomlRightPadded], p: P) -> Optional[TomlRightPadded]:
        if right is None:
            return None

        t = right.element
        if isinstance(t, Tree):
            new_t = self.visit(t, p)
        else:
            new_t = t

        after = self.visit_space(right.after, p)
        if after is right.after and new_t is right.element:
            return right
        return TomlRightPadded(new_t, after, right.markers)
